internal/keepers/filter.go

```diff
@@ -0,0 +1,176 @@
+// filter
+func (rc *reportCoordinator) Filter() func(types.UpkeepKey) bool {
+	return func(key types.UpkeepKey) bool {
+		id, err := rc.registry.IdentifierFromKey(key)
+		if err != nil {
+			return false
+		}
+
+		if _, ok := rc.idBlocks.Get(string(id)); ok {
+			return false
+		}
+
+		return true
+	}
+}
+
+func (rc *reportCoordinator) Accept(key types.UpkeepKey) {
+	rc.activeKeys.Set(string(key), false, defaultExpiration)
+}
+
+func (rc *reportCoordinator) IsTransmitting(key types.UpkeepKey) bool {
+	// key is assumed to be transmitted if it doesn't exist in cache or if it
+	// is no longer waiting for a transaction to be transmitted
+	inLogs, ok := rc.activeKeys.Get(string(key))
+	return ok && inLogs
+}
+
+func (rc *reportCoordinator) start() {
+	rc.starter.Do(func() {
+		go rc.run()
+		go rc.cacheCleaner.Run(rc.idBlocks)
+		go rc.cacheCleaner.Run(rc.activeKeys)
+	})
+}
+
+func (rc *reportCoordinator) stop() {
+	rc.chStop <- struct{}{}
+	rc.cacheCleaner.stop <- struct{}{}
+	rc.cacheCleaner.stop <- struct{}{}
+}
+
+func (rc *reportCoordinator) run() {
+	// TODO: handle panics by restarting this process
+
+	cadence := time.Second
+	timer := time.NewTimer(cadence)
+
+	for {
+
+		select {
+		case <-timer.C:
+			startTime := time.Now()
+			keys := rc.activeKeys.Keys()
+			logs, _ := rc.logs.PerformLogs(context.Background())
+
+			// log entries indicate that a perform exists on chain in some
+			// capacity. the existance of an entry means that the transaction
+			// was broadcast by at least one node. reorgs can still happen
+			// causing performs to vanish or get moved to a later block.
+			//
+			// in the case of reorgs causing a perform to be dropped from the
+			// chain, total confirmations are checked to allow other nodes
+			// to retry a transaction based on the OCR transmit cadence.
+			logLookup := make(map[string]types.PerformLog)
+			var minConfirmations int64 = 20
+			for _, log := range logs {
+				logLookup[string(log.Key)] = log
+				if _, ok := rc.activeKeys.Get(string(log.Key)); ok {
+					if log.Confirmations < minConfirmations {
+						// set the active key value to false to indicate that
+						// we now have a transaction identified and no new
+						// attempts at a transaction should be made
+						rc.activeKeys.Set(string(log.Key), true, defaultExpiration)
+					} else {
+						// the transaction is complete so the key can be deleted
+						rc.activeKeys.Delete(string(log.Key))
+					}
+				}
+
+				// if we detect a log, remove it from the observation filters
+				// to allow it to be reported on again
+				id, err := rc.registry.IdentifierFromKey(log.Key)
+				if err != nil {
+					continue
+				}
+
+				rc.idBlocks.Delete(string(id))
+			}
+
+			// perform logs can potentially come and go. for all keys, if a
+			// perform log is not detected, set the active keys value to true to
+			// indicate that we are waiting for a transaction log and any node
+			// should attempt to transmit.
```

Comment:
nit. true/false is opposite as of comment.

also not sure if it should be "any node should not attempt to transmit."?

Reply:
Damn. yes it is. I switched it and changed the function name because the former was confusing.